diff --binary -ur a/vendor/raylib/src/config.h b/vendor/raylib/src/config.h
--- a/vendor/raylib/src/config.h
+++ b/vendor/raylib/src/config.h
@@ -49,7 +49,7 @@
 // Use busy wait loop for timing sync, if not defined, a high-resolution timer is setup and used
 //#define SUPPORT_BUSY_WAIT_LOOP      1
 // Use a half-busy wait loop, in this case frame sleeps for some time and runs a busy-wait-loop at the end
-#define SUPPORT_HALFBUSY_WAIT_LOOP
+//#define SUPPORT_HALFBUSY_WAIT_LOOP
 // Wait for events passively (sleeping while no events) instead of polling them actively every frame
 //#define SUPPORT_EVENTS_WAITING      1
 // Allow automatic screen capture of current screen pressing F12, defined in KeyCallback()
